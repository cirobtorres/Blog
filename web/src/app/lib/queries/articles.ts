const GET_ARTICLES = `
query Articles($sort: [String], $filters: ArticleFiltersInput, $pagination: PaginationArg) {
  articles(sort: $sort, filters: $filters, pagination: $pagination) {
    documentId
    title
    slug
    description
    createdAt
    updatedAt
    publishedAt
    cover {
      documentId
      url
      alternativeText
      caption
      width
      height
    }
    category {
      documentId
      name
    }
    subCategories {
      documentId
      name
      link
      svg
    }
    tags {
      documentId
      name
    }
    author {
      name
      avatar {
        documentId
        url
        alternativeText
      }
    }
  }
}`;

const GET_ARTICLE = `
query Blocks($documentId: ID!) {
  article(documentId: $documentId) {
    documentId
    title
    slug
    description
    createdAt
    updatedAt
    publishedAt
    cover {
      documentId
      url
      alternativeText
      caption
      width
      height
    }
    author {
      name
      avatar {
        url
        alternativeText
      }
    }
    category {
      documentId
      name
    }
    subCategories {
      documentId
      name
      link
      svg
    }
    tags {
      documentId
      name
    }
    blocks {
      ... on ComponentSharedRichText {
        __typename
        id
        body
      }
      ... on ComponentSharedDetails {
        __typename
        id
        collapsible
        title
        svg
        body
      }
      ... on ComponentSharedQuiz {
        __typename
        id
        json
      }
      ... on ComponentSharedSlider {
        __typename
        id
        files {
          documentId
          url
          alternativeText
          caption
          width
          height
        }
      }
    }
  }
}`;

const COUNT_ARTICLES = `
query($filters: ArticleFiltersInput) {
  articles_connection(filters: $filters) {
    __typename
    pageInfo {
      __typename
      page
      pageCount
      pageSize
      total
    }
  }
}`;

export { GET_ARTICLES, GET_ARTICLE, COUNT_ARTICLES };
