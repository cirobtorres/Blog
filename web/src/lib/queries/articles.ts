const GET_ARTICLES = `
query Articles($sort: [String], $pagination: PaginationArg) {
  articles(sort: $sort, pagination: $pagination) {
    documentId
    title
    slug
    description
    createdAt
    updatedAt
    publishedAt
    cover {
      documentId
      url
      alternativeText
    }
    category {
      documentId
      name
    }
    author {
      name
      avatar {
        documentId
        url
        alternativeText
      }
    }
  }
}`;

const GET_ARTICLE = `
query Blocks($documentId: ID!) {
  article(documentId: $documentId) {
    documentId
    title
    slug
    description
    createdAt
    updatedAt
    publishedAt
    cover {
      url
      alternativeText
      caption
    }
    author {
      name
      avatar {
        url
        alternativeText
      }
    }
    category {
      documentId
      name
    }
    subCategories {
      documentId
      name
      link
      svg
    }
    tags {
      documentId
      name
    }
    blocks {
      ... on ComponentSharedRichText {
        __typename
        id
        body
      }
      ... on ComponentSharedQuote {
        __typename
        id
        body
        title
      }
      ... on ComponentSharedMedia {
        __typename
        id
        file {
          documentId
          url
          alternativeText
          caption
          width
          height
        }
      }
      ... on ComponentSharedSlider {
        __typename
        id
        files {
          documentId
          url
          alternativeText
          caption
          width
          height
        }
      }
    }
  }
}`;

export { GET_ARTICLES, GET_ARTICLE };
