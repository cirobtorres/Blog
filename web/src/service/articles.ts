import {
  graphqlReadArticleClient,
  graphqlLikeArticleClient,
} from "../lib/graphQlClient";
import {
  COUNT_ARTICLES,
  COUNT_LIKES,
  DISLIKE_ARTICLE,
  GET_ARTICLE,
  GET_ARTICLES,
  LIKE_ARTICLE,
} from "../lib/queries/articles";

const getArticles = async (
  sort?: string | null,
  sortBy?: string | null,
  pagination: {
    page?: number | null;
    pageSize?: number | null;
    start?: number | null;
    limit?: number | null;
  } = {
    page: null,
    pageSize: null,
    start: null,
    limit: null,
  }
) => {
  try {
    const searchTerms = sortBy
      ? sortBy.split("-").filter((term) => term.trim() !== "")
      : null;

    const filters = {
      and: searchTerms?.map((term) => ({
        or: [
          { slug: { contains: term } },
          { topic: { slug: { contains: term } } },
          { tools: { slug: { contains: term } } },
          { tags: { slug: { contains: term } } },
        ],
      })),
    };

    const { articles }: { articles: ArticleCard[] } =
      await graphqlReadArticleClient.request(GET_ARTICLES, {
        sort,
        filters,
        pagination,
      });

    return { data: articles };
  } catch (error) {
    console.error("Failed to fetch articles:", error);
    throw new Error("Failed to fetch articles");
  }
};

const getArticle = async (documentId: string) => {
  try {
    const { article }: { article: Article } =
      await graphqlReadArticleClient.request(GET_ARTICLE, {
        documentId,
      });

    return { data: article };
  } catch (error) {
    console.error("Failed to fetch article:", error);
    throw new Error("Failed to fetch article");
  }
};

const countArticles = async (sortBy?: string | null) => {
  try {
    const searchTerms = sortBy
      ? sortBy.split("-").filter((term) => term.trim() !== "")
      : null;

    const filters = {
      and: searchTerms?.map((term) => ({
        or: [
          { slug: { contains: term } },
          { topic: { slug: { contains: term } } },
          { tools: { slug: { contains: term } } },
          { tags: { slug: { contains: term } } },
        ],
      })),
    };

    const {
      articles_connection: { pageInfo },
    }: {
      articles_connection: {
        __typename: string;
        pageInfo: {
          __typename: string;
          page: number;
          pageCount: number;
          pageSize: number;
          total: number;
        };
      };
    } = await graphqlReadArticleClient.request(COUNT_ARTICLES, {
      filters,
    });

    return { data: pageInfo };
  } catch (error) {
    console.error("Failed to fetch count articles:", error);
    throw new Error("Failed to fetch count articles");
  }
};

const countArticleLikes = async (articleDocumentId: string) => {
  try {
    const filters = {
      article: {
        documentId: {
          eq: articleDocumentId,
        },
      },
    };

    const {
      articleLikes,
    }: {
      articleLikes: {
        documentId: string;
        users_permissions_user: {
          documentId: string;
        };
      }[];
    } = await graphqlLikeArticleClient.request(COUNT_LIKES, {
      filters,
    });

    return { data: articleLikes };
  } catch (error) {
    console.error("Failed to count article likes:", error);
    throw new Error("Failed to count article likes");
  }
};

const likeArticle = async (
  articleDocumentId: string,
  userDocumentId: string
) => {
  try {
    const data = {
      article: articleDocumentId,
      users_permissions_user: userDocumentId,
    };

    const {
      createArticleLike: { documentId },
    }: {
      createArticleLike: {
        documentId: string;
        users_permissions_user: string;
      };
    } = await graphqlLikeArticleClient.request(LIKE_ARTICLE, {
      documentId: articleDocumentId,
      data,
    });

    return { data: documentId };
  } catch (error) {
    console.error("Failed to like article:", error);
    throw new Error("Failed to like article");
  }
};

const dislikeArticle = async (articleDocumentId: string) => {
  try {
    const {
      deleteArticleLike: { documentId },
    }: {
      deleteArticleLike: {
        documentId: string;
      };
    } = await graphqlLikeArticleClient.request(DISLIKE_ARTICLE, {
      documentId: articleDocumentId,
    });

    return { data: documentId };
  } catch (error) {
    console.error("Failed to dislike article:", error);
    throw new Error("Failed to dislike article");
  }
};

export {
  getArticles,
  getArticle,
  countArticles,
  countArticleLikes,
  likeArticle,
  dislikeArticle,
};
