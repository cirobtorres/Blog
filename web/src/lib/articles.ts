import graphqlClient from "./graphQlClient";
import { GET_ARTICLE, GET_ARTICLES } from "./queries/articles";

const getArticles = async (
  sort?: string | null,
  pagination: {
    page?: number | null;
    pageSize?: number | null;
    start?: number | null;
    limit?: number | null;
  } = {
    page: null,
    pageSize: null,
    start: null,
    limit: null,
  }
) => {
  try {
    const { articles }: { articles: ArticleCards } =
      await graphqlClient.request(GET_ARTICLES, {
        sort,
        pagination,
      });
    return { data: articles };
  } catch (error) {
    console.error("Failed to fetch articles:", error);
    throw new Error("Failed to fetch articles");
  }
};

const getArticle = async (documentId: string) => {
  try {
    const { article }: { article: Article } = await graphqlClient.request(
      GET_ARTICLE,
      {
        documentId,
      }
    );
    return { data: article };
  } catch (error) {
    console.error("Failed to fetch article:", error);
    throw new Error("Failed to fetch article");
  }
};

export { getArticles, getArticle };
