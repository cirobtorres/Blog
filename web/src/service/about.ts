import graphqlClient from "../lib/graphQlClient";
import { GET_ABOUT } from "../lib/queries/about";

const getAbout = async () => {
  try {
    const { about }: { about: About } = await graphqlClient.request(GET_ABOUT);
    return { data: about };
  } catch (error) {
    console.error("Failed to fetch about:", error);
    throw new Error("Failed to fetch about");
  }
};

export { getAbout };
