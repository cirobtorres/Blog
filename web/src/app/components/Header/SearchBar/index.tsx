"use client";

import { Search } from "lucide-react";
import { useDebouncedCallback } from "use-debounce";
import { ReadonlyURLSearchParams, useSearchParams } from "next/navigation";
import { useState } from "react";
import slugify from "../../../utils/slugify";
import { getArticles } from "../../../lib/articles";
import Link from "next/link";
import Image from "next/image";
import {
  Dialog,
  DialogClose,
  DialogContent,
  DialogDescription,
  DialogHeader,
  DialogTitle,
  DialogTrigger,
} from "../../Shadcnui/dialog";

type ArticleList = {
  documentId: string;
  title: string;
  slug: string;
  cover: {
    documentId: string;
    url: string;
    alternativeText: string;
  };
}[];

const SearchBar = () => {
  const searchParams = useSearchParams();
  const [sortedList, setSortedList] = useState<ArticleList | null>(null);

  const handleSearch = useDebouncedCallback(async (queryText: string) => {
    const params = new URLSearchParams(searchParams);

    // Delete user's input if it is empty
    if (queryText) {
      params.set("query", queryText);
    } else {
      params.delete("query");
    }

    const sortBySlugfied = slugify(params.get("query") || "");
    if (sortBySlugfied !== "") {
      const { data: articleList } = await getArticles(
        "createdAt:desc",
        sortBySlugfied
      );

      setSortedList(articleList.length === 0 ? [] : articleList.slice(0, 7));
    } else {
      setSortedList(null);
    }
  }, 500);

  return (
    <Dialog>
      <DialogTrigger
        className="w-full rounded-3xl flex-1" // outline-none
      >
        <Trigger />
      </DialogTrigger>
      <DialogContent>
        <DialogHeader>
          <DialogTitle className="sr-only">Buscador de artigos</DialogTitle>
        </DialogHeader>
        <form action="/artigos" className="overflow-x-hidden">
          <InputQuery handleSearch={handleSearch} searchParams={searchParams} />
          <ResultQuery sortedList={sortedList} />
        </form>
      </DialogContent>
    </Dialog>
  );
};

const Trigger = () => {
  return (
    <div className="relative w-full min-h-10 flex items-center rounded-3xl border border-blog-border bg-blog-background-1 transition-all duration-500 hover:bg-blog-background-2 group">
      <Search className="absolute left-3 top-1/2 -translate-y-1/2 size-4 transition-all duration-500 text-blog-foreground-readable group-hover:text-blog-foreground-readable-hover" />
      <p className="pl-12 w-full text-sm text-start transition-all duration-500 group-hover:text-blog-foreground-readable-hover">
        Buscar artigos...
      </p>
    </div>
  );
};

const InputQuery = ({
  searchParams,
  handleSearch,
}: {
  searchParams: ReadonlyURLSearchParams;
  handleSearch: (queryString: string) => void;
}) => {
  return (
    <div className="relative min-h-10 w-full py-4">
      <div className="flex flex-col justify-center">
        <button
          type="submit"
          className="absolute left-4 top-1/2 -translate-y-1/2"
        >
          <Search className="size-4 transition-all duration-500 text-blog-foreground-readable hover:text-blog-foreground-readable-hover" />
        </button>
        <input
          id="query"
          type="query"
          name="query"
          autoComplete="off"
          placeholder="Pesquisar artigos..."
          onChange={(event) => {
            handleSearch(event.target.value);
          }}
          defaultValue={searchParams.get("query")?.toString()}
          className="text-sm w-full border-transparent border-none border-0 outline-none outline-0 pl-12 pr-16 bg-transparent"
        />
      </div>
      <DialogClose className="absolute right-4 top-1/2 -translate-y-1/2 px-2 py-0.5 text-xs font-extrabold rounded border border-blog-border transition-all duration-500 bg-blog-background-2 hover:text-blog-foreground-readable-hover hover:bg-blog-background-1">
        <DialogDescription className="sr-only">
          Fechar buscador
        </DialogDescription>
        ESC
      </DialogClose>
    </div>
  );
};

const ResultQuery = ({ sortedList }: { sortedList: ArticleList | null }) => {
  return (
    <>
      {sortedList && sortedList.length > 0 && (
        <>
          <hr />
          <ul className="flex flex-col py-4">
            {sortedList.map((article) => (
              <li key={article.documentId}>
                <Link
                  href={`/artigos/[documentId]/[slug]`}
                  as={`/artigos/${article.documentId}/${article.slug}`}
                  className="flex items-center gap-3 py-1.5 px-5"
                >
                  <div className="shrink-0 relative size-8">
                    <Image
                      src={`http://127.0.0.1:1337${article.cover.url}`}
                      alt={article.cover.alternativeText}
                      fill
                      sizes="(min-width: 16px) 100vw"
                      className="absolute object-cover"
                    />
                  </div>
                  <p className="text-sm truncate text-blog-foreground-readable">
                    {article.title}
                  </p>
                </Link>
              </li>
            ))}
          </ul>
        </>
      )}
      {sortedList && sortedList.length === 0 && (
        <>
          <hr />
          <div className="py-4">
            <span className="sr-only">Nenhum artigo encontrado</span>
            <p className="text-center text-sm text-blog-foreground-readable">
              Nenhum artigo encontrado
            </p>
          </div>
        </>
      )}
    </>
  );
};

export default SearchBar;
