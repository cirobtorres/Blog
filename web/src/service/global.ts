import { graphqlReadGlobalClient } from "../lib/graphQlClient";
import { GET_GLOBAL } from "../lib/queries/global";

const getGlobal = async () => {
  try {
    const { global }: { global: Global } =
      await graphqlReadGlobalClient.request(GET_GLOBAL);
    return { data: global };
  } catch (error) {
    console.error("Failed to fetch global:", error);
    throw new Error("Failed to fetch global");
  }
};

export { getGlobal };
