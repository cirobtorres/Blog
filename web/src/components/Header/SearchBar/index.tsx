"use client";

import { Search } from "lucide-react";
import { useDebouncedCallback } from "use-debounce";
import { ReadonlyURLSearchParams, useSearchParams } from "next/navigation";
import { useState } from "react";
import { getArticles } from "../../../service/articles";
import slugify from "../../../utils/slugify";
import Link from "next/link";
import Image from "next/image";
import {
  Dialog,
  DialogClose,
  DialogContent,
  DialogDescription,
  DialogHeader,
  DialogTitle,
  DialogTrigger,
} from "../../Shadcnui/dialog";

type ArticleList = {
  documentId: string;
  title: string;
  slug: string;
  cover: {
    documentId: string;
    url: string;
    alternativeText: string;
  };
}[];

const SearchBar = () => {
  const searchParams = useSearchParams();
  const [sortedList, setSortedList] = useState<ArticleList | null>(null);

  const handleSearch = useDebouncedCallback(async (queryText: string) => {
    const params = new URLSearchParams(searchParams);

    // Delete user's input if it is empty
    if (queryText) {
      params.set("query", queryText);
    } else {
      params.delete("query");
    }

    const sortBySlugfied = slugify(params.get("query") || "");
    if (sortBySlugfied !== "") {
      const { data: articleList } = await getArticles(
        "createdAt:desc",
        sortBySlugfied
      );

      setSortedList(articleList.length === 0 ? [] : articleList.slice(0, 7));
    } else {
      setSortedList(null);
    }
  }, 500);

  return (
    <Dialog>
      <DialogTrigger
        data-testid="search-bar-trigger-button"
        className="w-full rounded flex-1 focus-visible:outline focus-visible:outline-2 focus-visible:outline-blog-foreground-readable-hover"
      >
        <Trigger />
      </DialogTrigger>
      <DialogContent>
        <DialogHeader>
          <DialogTitle className="sr-only">Buscador de artigos</DialogTitle>
        </DialogHeader>
        <form action="/artigos" className="overflow-x-hidden">
          <InputQuery
            data-testid="search-bar-input-query"
            handleSearch={handleSearch}
            searchParams={searchParams}
          />
          <ResultQuery sortedList={sortedList} />
        </form>
      </DialogContent>
    </Dialog>
  );
};

const Trigger = () => {
  // The search bar that displays on header content
  return (
    <div
      data-testid="search-bar-trigger"
      className="relative w-full min-h-9 flex items-center rounded border border-blog-border bg-[#1d1d1d] transition-all duration-500 hover:bg-blog-border group"
    >
      <Search className="absolute left-3 top-1/2 -translate-y-1/2 size-4 transition-all duration-500 text-blog-foreground-readable group-hover:text-blog-foreground-readable-hover" />
      <p className="pl-12 w-full text-sm text-start transition-all duration-500 group-hover:text-blog-foreground-readable-hover">
        Buscar artigos...
      </p>
    </div>
  );
};

const InputQuery = ({
  searchParams,
  handleSearch,
}: {
  searchParams: ReadonlyURLSearchParams;
  handleSearch: (queryString: string) => void;
}) => {
  // This is the search bar once the user has clicked it (that appears inside the dialog)
  return (
    <div className="relative min-h-10 w-full py-4">
      <div className="flex flex-col justify-center">
        <button
          data-testid="search-bar-submit"
          type="submit"
          className="absolute left-4 top-1/2 -translate-y-1/2"
        >
          <Search className="size-4 transition-all duration-500 text-blog-foreground-readable hover:text-blog-foreground-readable-hover" />
        </button>
        <input
          id="search-bar-query"
          data-testid="search-bar-query"
          type="query"
          name="query"
          autoComplete="off"
          placeholder="Pesquisar artigos..."
          autoFocus
          onChange={(event) => {
            handleSearch(event.target.value);
          }}
          defaultValue={searchParams.get("query")?.toString()}
          className="text-sm w-full border-transparent border-none border-0 outline-none outline-0 pl-12 pr-16 bg-transparent"
        />
      </div>
      <DialogClose
        data-testid="search-bar-close"
        className="absolute right-4 top-1/2 -translate-y-1/2 px-2 py-0.5 text-xs font-extrabold rounded border border-blog-border transition-all duration-500 bg-blog-background-2 hover:text-blog-foreground-readable-hover hover:bg-blog-background-1"
      >
        <DialogDescription className="sr-only">
          Fechar buscador
        </DialogDescription>
        ESC
      </DialogClose>
    </div>
  );
};

const ResultQuery = ({ sortedList }: { sortedList: ArticleList | null }) => {
  // This is the search result box that appears bellow the search bar
  return (
    <>
      {sortedList && sortedList.length > 0 && (
        <>
          <hr />
          <ul className="flex flex-col py-4">
            {sortedList.map((article) => (
              <li key={article.documentId}>
                <Link
                  href={`/artigos/[documentId]/[slug]`}
                  as={`/artigos/${article.documentId}/${article.slug}`}
                  className="flex items-center gap-3 py-1.5 px-5"
                >
                  <div className="shrink-0 relative size-8">
                    <Image
                      src={`http://127.0.0.1:1337${article.cover.url}`}
                      alt={article.cover.alternativeText}
                      fill
                      sizes="(min-width: 16px) 100vw"
                      className="absolute object-cover"
                    />
                  </div>
                  <p className="text-sm truncate text-blog-foreground-readable">
                    {article.title}
                  </p>
                </Link>
              </li>
            ))}
          </ul>
        </>
      )}
      {sortedList && sortedList.length === 0 && (
        <>
          <hr />
          <div data-testid="search-bar-none-found" className="py-4">
            <span className="sr-only">Nenhum artigo encontrado</span>
            <p className="text-center text-sm text-blog-foreground-readable">
              Nenhum artigo encontrado
            </p>
          </div>
        </>
      )}
    </>
  );
};

export default SearchBar;
