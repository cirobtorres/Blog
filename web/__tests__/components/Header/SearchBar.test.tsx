import "@testing-library/jest-dom";
import SearchBar from "../../../src/components/Header/SearchBar";
import {
  render,
  screen,
  fireEvent,
  waitFor,
  act,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { getArticles } from "../../../src/service/articles";
import { useSearchParams } from "next/navigation";

jest.mock("../../../src/service/articles");

jest.mock("next/navigation", () => ({
  useSearchParams: jest.fn(),
}));

describe("SearchBar", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders the search bar component", () => {
    render(<SearchBar />);
    const searchBar = screen.getByTestId("search-bar-trigger");
    expect(searchBar).toBeInTheDocument();
  });

  it("does not render input query dialog box", () => {
    render(<SearchBar />);
    const submitButton = screen.queryByTestId("search-bar-input-query");
    expect(submitButton).not.toBeInTheDocument();
  });

  it("clicks DialogTrigger to open InputQuery and searches article on ResultQuery", async () => {
    (getArticles as jest.Mock).mockResolvedValue({
      data: [
        {
          documentId: "1",
          title: "Article 1",
          slug: "article-1",
          cover: {
            documentId: "1",
            url: "/path/to/image1.jpg",
            alternativeText: "Article 1",
          },
        },
      ],
    });

    (useSearchParams as jest.Mock).mockReturnValue({
      get: jest.fn().mockReturnValue("article-1"),
    });

    render(<SearchBar />);

    const buttonTrigger = screen.getByTestId("search-bar-trigger-button");
    fireEvent.click(buttonTrigger);
    await waitFor(() => expect(screen.getByRole("dialog")).toBeInTheDocument());

    const inputQuery = await waitFor(() =>
      screen.getByTestId("search-bar-query")
    );
    expect(inputQuery).toBeInTheDocument();

    await waitFor(() => userEvent.type(inputQuery, "Article 1"));
    await waitFor(() =>
      expect(screen.getByText("Article 1")).toBeInTheDocument()
    );
  });

  it("displays 'No articles found' message when no results are returned", async () => {
    (getArticles as jest.Mock).mockResolvedValue({ data: [] }); // Mock: no article on database

    render(<SearchBar />);

    // Clicks the trigger and await for it to open the dialog box
    const buttonTrigger = screen.getByTestId("search-bar-trigger-button");
    fireEvent.click(buttonTrigger);
    await waitFor(() => expect(screen.getByRole("dialog")).toBeInTheDocument());

    // Clears the input, just to make sure, and type in "Article X"
    const inputQuery = await waitFor(() =>
      screen.getByTestId("search-bar-query")
    );
    await waitFor(() => userEvent.clear(inputQuery));
    await waitFor(() => userEvent.type(inputQuery, "Article X"));

    // Await for server to respond its call
    await waitFor(() =>
      expect(getArticles).toHaveBeenCalledWith("createdAt:desc", "article-x")
    );

    // Make assertion
    const inputField = await waitFor(() =>
      screen.getByTestId("search-bar-none-found")
    );
    expect(inputField).toHaveTextContent("Nenhum artigo encontrado");
  });

  it("calls the handleSearch function with debounced value", async () => {
    jest.useFakeTimers();
    (getArticles as jest.Mock).mockResolvedValue({
      data: [
        {
          documentId: "2",
          title: "Article 2",
          slug: "article-2",
          cover: {
            documentId: "2",
            url: "/path/to/image2.jpg",
            alternativeText: "Article 2",
          },
        },
      ],
    });

    render(<SearchBar />);

    const buttonTrigger = screen.getByTestId("search-bar-trigger-button");
    fireEvent.click(buttonTrigger);
    await waitFor(() => expect(screen.getByRole("dialog")).toBeInTheDocument());

    const inputQuery = screen.getByTestId("search-bar-query");

    await act(async () => {
      userEvent.clear(inputQuery);
      userEvent.type(inputQuery, "Article 2");
      jest.advanceTimersByTime(500); // Advance time according to useDebouncedCallback
    });

    await waitFor(() => expect(inputQuery).toHaveValue("Article 2")); // Guarantees that inputQuery has "Article 2" as text input before following to next assertions
    await waitFor(() => expect(getArticles).toHaveBeenCalled());
    await waitFor(() =>
      expect(getArticles).toHaveBeenCalledWith("createdAt:desc", "article-2")
    );

    jest.useRealTimers();
  });

  it("closes the dialog when ESC is pressed", async () => {
    render(<SearchBar />);

    const buttonTrigger = screen.getByTestId("search-bar-trigger-button");
    fireEvent.click(buttonTrigger);
    await waitFor(() => expect(screen.getByRole("dialog")).toBeInTheDocument());

    const closeButton = screen.getByTestId("search-bar-close");
    fireEvent.click(closeButton);
    expect(screen.queryByTestId("search-bar-trigger")).toBeInTheDocument();
  });

  it("matches the snapshot", () => {
    const { asFragment } = render(<SearchBar />);
    expect(asFragment()).toMatchSnapshot();
  });
});
