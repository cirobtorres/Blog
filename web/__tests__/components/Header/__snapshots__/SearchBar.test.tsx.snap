// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar matches the snapshot 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-:ri:"
    aria-expanded="false"
    aria-haspopup="dialog"
    class="w-full rounded flex-1 focus-visible:outline focus-visible:outline-2 focus-visible:outline-blog-foreground-readable-hover group"
    data-state="closed"
    data-testid="search-bar-trigger-button"
    type="button"
  >
    <div
      class="relative w-full min-h-9 flex items-center rounded bg-[#1d1d1d] bg-opacity-75 hover:bg-blog-border group-focus-visible:bg-blog-border transition-all duration-500 group"
      data-testid="search-bar-trigger"
    >
      <svg
        class="lucide lucide-search absolute left-3 top-1/2 -translate-y-1/2 size-4 transition-all duration-500 text-blog-foreground-readable group-hover:text-blog-foreground-readable-hover"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="11"
          cy="11"
          r="8"
        />
        <path
          d="m21 21-4.3-4.3"
        />
      </svg>
      <p
        class="pl-12 w-full text-sm text-start transition-all duration-500 group-hover:text-blog-foreground-readable-hover"
      >
        Buscar artigos...
      </p>
    </div>
  </button>
</DocumentFragment>
`;
